### ISSUE TITLE
add precedence for "blockers" in todo() method

### ISSUE THREAD

**Issue by capsulecorplab** [capsulecorplab/mindgraph#20]
Current `todo()` method implementation, only accounts for `Task.subtasks` and `Task.priority`. Needs to also account for `Task.blockers`, such that `Task` objects in "blockers" yield prior to `Task` objects in "subtasks".

**Comment by yamad** [comment]
Can you clarify the distinction between what is supposed to be in `threads` vs `dependencies`?

While writing the `todo` method, I interpreted everything in `threads` to be items that had to be completed before the parent node could complete.  That is, a `thread` _is_ a dependency. At least that is what I understood from the example.

**Comment by capsulecorplab** [comment]
`Node.dependencies` was intended as a list of nodes that should precede `Node.threads`, at least in term of yield order; `Node.dependencies` can be thought of as an extension of `Node.threads` with higher weights, as far as the `todo` method is concerned.

Also, I'm thinking the following changes to the current naming convention might help:
`Graph` to `Project`
`Node` to `Task`
`Node.threads` to `Task.subtasks`
`Node.dependencies` to `Task.blockers`
`Node.weight` to `Task.priority`

Thoughts?

**Comment by yamad** [comment]
Yes, I like those name changes a lot. Much clearer. That makes the required change to `todo` pretty clear too.

So the distinction between a subtask and a blocker is that sometimes you'll want to display just one or the other?

While we are at a name overhaul, I'm not convinced that `Graph`/`Project` is paying its way as a separate object at all. Maybe it could just alias `Task` like `Project = Task`?

**Comment by capsulecorplab** [comment]
> Yes, I like those name changes a lot. Much clearer. That makes the required change to todo pretty clear too.

Cool! Will put up a separate issue just for the name changes.

> So the distinction between a subtask and a blocker is that sometimes you'll want to display just one or the other?

Not sure what you mean. As far as the `task.__repr__()` or `print(task)` methods are concerned, I think they should both be displayed, but I haven't quite figured out the best way to display the difference; maybe an  `->` symbol to indicate blocked by? e.g.,
```
>>> myproject
Build cool stuff:
    - invent a flux capacitor
    - build a time machine out of a DeLorean -> invent a flux capacitor
        - retrofit a DeLorean
        - add time circuits
        - add a flux capacitor
        - add a fusion reactor that generates 1.21 gigawatts
    - build a dog feeder
```
^This might have to be its own issue.

> While we are at a name overhaul, I'm not convinced that Graph/Project is paying its way as a separate object at all. Maybe it could just alias Task like Project = Task?

The only difference, class-wise, is that `Project `can output to a YAML file. Otherwise, you're right, project might as well be aliased as a task.

**Comment by yamad** [comment]
> Not sure what you mean.  ...

Sorry, I was trying to understand better the justification for keeping two independent classes of blocking tasks. How do they get treated differently in each method? For `todo`, as you say, the connection between the two types is pretty close.

The downside, if you can call it that, of having two independent edge types is that the connection to graph algorithms becomes less clear. For example, `_postorder` becomes a bad name for a method that treats two types of edges completely differently.

> The only difference, class-wise, is that Project can output to a YAML file.

Sure, but why not allow any Task to be output to a yaml file?

Probably makes sense to do the name changes first, before editing the `todo` method. Also, someone else should feel free to make the necessary changes.

**Comment by capsulecorplab** [comment]
Open to counter-arguments, but in principle, I feel like subtasks and its parent task should share a composition relationship, whereas blockers and the task it's blocking should share an aggregation relationship - i.e., subtasks can't be shared by multiple tasks, whereas a blocker is a task that exists somewhere else within the project tree. 
Not sure what the best way to enforce these relationship constraints in practice, other than to run a search every time the `append`, `blockedby`, or `blocking` methods are called.
As for the functionality of `todo`, post-order traversal can still work so long as it yields blockers first, followed by subtasks, while still sorting by priority, if that makes sense.

> Sure, but why not allow any Task to be output to a yaml file?

I'm open to that change. In which case, it might also make sense to refactor `Project` into a factory method that constructs and returns a `Task` object, just so it feels like a user is "creating a project".
